### PR TITLE
[do not merge] Gmaps revival

### LINKED ISSUE #3289: use user api key for embeds/public pages with google maps enabled
depends on #3025

### PR DESCRIPTION
**Pull request by javisantana**
Let's get back GMaps!
This is not ready but it's the base to add the api key stuff.

### To make it work:

In order to have google maps you need to:

1. Add the following to the basemaps section in app_config.yml

```
GMaps:
      roadmap:
        name: 'GMaps Roadmap'
        maxZoom: 40
        minZoom: 0
        baseType: 'roadmap'
        className: 'googlemaps'
      hybrid:
        name: 'GMaps Hybrid'
        maxZoom: 40
        minZoom: 0
        baseType: 'hybrid'
        className: 'googlemaps'
      satellite:
        name: 'GMaps Satellite'
        maxZoom: 40
        minZoom: 0
        baseType: 'satellite'
        className: 'googlemaps'
      gray_roadmap:
        name: 'GMaps Gray Roadmap'
        maxZoom: 40
        minZoom: 0
        baseType: 'gray_roadmap'
        className: 'googlemaps'
        style: '[{ "stylers": [ { "saturation": -100 } ] },{ "featureType": "water", "stylers": [ { "gamma": 1.67 }, { "lightness": 27 } ] },{ "elementType": "geometry", "stylers": [ { "gamma": 1.31 }, { "lightness": 12 } ] },{ "featureType": "administrative", "elementType": "labels", "stylers": [ { "lightness": 51 }, { "gamma": 0.94 } ] },{ },{ "featureType": "road", "elementType": "labels", "stylers": [ { "lightness": 57 } ] },{ "featureType": "poi", "elementType": "labels", "stylers": [ { "lightness": 42 } ] }]'
```

2. Enable feature flag ```google_maps```.

---

Fixes #3289.

### CODE REVIEW

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3222/))

File: lib/assets/javascripts/cartodb/table/mapview.js [hunk 1]
```diff
@@ -400,9 +400,13 @@ cdb.admin.MapTab = cdb.core.View.extend({
     var baseLayer = this.map.getBaseLayer();
     var provider  = this.map.get("provider");
 
-    if (baseLayer && provider && provider.toLowerCase().indexOf("googlemaps") !== -1) {
-      this._showGMapsDeprecationDialog();
-      return;
+    if (!this.user.featureEnabled('google')) {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/table/mapview.js]: I'd use gmaps or google-maps, because google has lot of services.

Comment (javisantana) [inline lib/assets/javascripts/cartodb/table/mapview.js]: I'm going to remove this since having google maps depends on configuration, sounds good?

Comment (xavijam) [inline lib/assets/javascripts/cartodb/table/mapview.js]: Better!

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3231/))

**Comment by javisantana** [comment]
cc @xavijam enable "google_maps" feature flag

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3232/))

File: app/controllers/admin/organizations_controller.rb [hunk 2]
```diff
@@ -39,6 +39,9 @@ def settings_update
     @organization.display_name = attributes[:display_name]
     @organization.discus_shortname = attributes[:discus_shortname]
     @organization.twitter_username = attributes[:twitter_username]
+    if current_user.has_feature_flag?('google_maps') && attributes.include?(:google_maps_key)
```
Comment (xavijam) [inline app/controllers/admin/organizations_controller.rb]: For simplifying things and not create two different feature flags for the same stuff, only it is necessary to have the organization owner with the feature flag enabled.

cc @Kartones @javisantana

File: app/views/admin/users/account.html.erb [hunk 3]
```diff
@@ -102,6 +102,31 @@
...
+              <% if current_user.organization.present? %>
+                <input type="text" value="<%= current_user.organization.google_maps_key %>" class="Form-input Form-input--long is-disabled" placeholder="<%= 'Add key in organization settings' if current_user.organization_owner? %>" readonly />
+              <% else %>
+                <%= f.text_field :google_maps_key, class: "Form-input Form-input--long #{ 'Form-input--error' if @user.errors[:gmaps].present? }", placeholder: "Copy here your API key" %>
+              <% end %>
+            </div>
+            <% if current_user.organization.present? %>
+              <div class="Form-rowInfo">
+                <p class="Form-rowInfoText">This is the organization Google Maps key</p>
+              </div>
```
Comment (xavijam) [inline app/views/admin/users/account.html.erb]: If the user belongs to an organization, and he has the feature flag enabled, a text saying that 'organization google maps rules' will be displayed.

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3234/))

File: app/controllers/admin/organizations_controller.rb (hunk 2 above)
Comment (javisantana) [inline app/controllers/admin/organizations_controller.rb]: :es:

**Comment by xavijam** [comment]
retest this please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3237/))

File: app/controllers/admin/users_controller.rb [hunk 4]
```diff
@@ -47,6 +47,12 @@ def account_update
     if @user.can_change_email && attributes[:email].present?
       @user.set_fields(attributes, [:email])
     end
+
+    if attributes[:google_maps_key].present? && !@user.organization.present?
+      @user.set_fields(attributes, [:google_maps_key])
+    end
+
+    @user.set_fields(attributes, [:email])
```
Comment (Kartones) [inline app/controllers/admin/users_controller.rb]: Warning, why is line 55 there, if is already added before at line 47-49 ? looks like some typo

Comment (xavijam) [inline app/controllers/admin/users_controller.rb]: Thanks @Kartones

File: lib/assets/javascripts/cartodb/table/basemap/basemap_dropdown/basemap_dropdown.js [hunk 5]
```diff
@@ -113,7 +116,7 @@ cdb.admin.DropdownBasemap = cdb.ui.common.Dropdown.extend({
 
   add: function(lyr) {
 
-    if (!lyr.get("urlTemplate")) return;
+    //if (!lyr.get("urlTemplate")) return;
```
Comment (Kartones) [inline lib/assets/javascripts/cartodb/table/basemap/basemap_dropdown/basemap_dropdown.js]: Don't comment code, delete it! :grinning:

File: lib/assets/javascripts/cartodb/table/mapview.js [hunk 6]
```diff
@@ -479,6 +483,9 @@ cdb.admin.MapTab = cdb.core.View.extend({
     var div = this.$('.cartodb-map');
 
     var mapViewClass = cdb.admin.LeafletMapView;
+    if (this.map.get('provider') === 'googlemaps') {
```
Comment (Kartones) [inline lib/assets/javascripts/cartodb/table/mapview.js]: Why not ternary if? looks ugly to define twice `mapViewClass`, when you can just do 
`var mapViewClass = this.map.get('provider') === 'googlemaps' ? cdb.admin.GoogleMapsMapView : cdb.admin.LeafletMapView; `

**Comment by Kartones** [comment]
:+1:

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3249/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3357/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3361/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3364/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3366/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3370/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3371/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3372/))

**Comment by Kartones** [comment]
:+1:  looks good.

Tomorrow I can implement required changes for the refactor

**Pull request closed without merging**